File: LeetCode/Problems/0027_Remove_Element/remove_element.py
from typing import List

class Solution:
    def removeElement(self, nums: List[int], val: int) -> int:
        return self.__twoPointers(nums, val)
        # return self.__swapValues(nums, val)

    # time complexity: O(n), space complexity: O(1)
    def __twoPointers(self, nums: List[int], val: int) -> int:
        if len(nums) == 0:
            return 0

        i = 0
        for n in nums:
            if n != val:
                nums[i] = n
                i += 1
        return i

    # time complexity: O(n), space complexity: O(1)
    def __swapValues(self, nums: List[int], val: int) -> int:
        if len(nums) == 0:
            return 0

        l = 0
        r = len(nums) - 1
        while l <= r:
            if nums[l] == val:
                # nums[l], nums[r] = nums[r], nums[l]
                nums[l] = nums[r]
                r -= 1
            else:
                l += 1
        return l

def printList(nums: List[int]) -> str:
    return (', ').join(map(str, nums))

nums = [3,2,2,3]
val = 3
print("Input: nums = [" + printList(nums) + "], val = " + str(val))

size = Solution().removeElement(nums, val)
print("Output: nums = [" + printList(nums[:size]) + "]\n")

nums = [0,1,2,2,3,0,4,2]
val = 2
print("Input: nums = [" + printList(nums) + "], val = " + str(val))

size = Solution().removeElement(nums, val)
print("Output: nums = [" + printList(nums[:size]) + "]")

File: LeetCode/Problems/0027_Remove_Element/remove_element.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int removeElement(vector<int>& nums, int val) {
        return twoPointers(nums, val);
        // return swapValues(nums, val);
    }

private:
    // time complexity: O(n), space complexity: O(1)
    int twoPointers(vector<int>& nums, int val){
        int i = 0;

        for(const auto& n : nums)
            if(n != val)
                nums[i++] = n;

        return i;
    }

    // time complexity: O(n), space complexity: O(1)
    int swapValues(vector<int>& nums, int val){
        int l = 0;
        int r = nums.size() - 1;

        while(l <= r){
            if(nums[l] == val)
                nums[l] = nums[r--];  // same as swap()
            else
                l++;
        }

        return l;
    }
};

// nums is passed by constant reference and cannot be changed by this function
string printVector(const vector<int>& nums, int len){
    string str;

    str.append("[");
    for(int i = 0; i < len; i++){
        if(i != 0)
            str.append(", ");
        str.append(std::to_string(nums[i]));
    }
    str.append("]");

    return str;
}

int main(){
    int val = 3;
    vector<int> nums({3, 2, 2, 3});
    cout << "Input: " << printVector(nums, 4) << "\n";

    Solution sol;
    int len = sol.removeElement(nums, val);
    cout << "Output: " << printVector(nums, len) << "\n";

    return 0;
}
